ops/opset13/lstm.go

```diff
@@ -0,0 +1,409 @@
+//
+//	o = f(Xt*(W^T) + Wb + H*(R^T) + Rb + P (.) C)
+//
+// Where:
+//   - 'f()' is an activation function
+//   - 'Xt' is the input tensor
+//   - 'W' is the input weight
+//   - 'Wb' is the input bias
+//   - 'H' is the hidden tensor
+//   - 'R' is the hidden weight tensor
+//   - 'Rb' is the hidden bias
+//   - 'P' are peephole weights (optional, can be nil)
+//   - 'C' is the cell state
+//   - '(.)' is element-wise multiplication
+//
+// 'o' is the result tensor that is returned.
+// This calculation can be used for the forget gate, input gate, cell gate
+// and output gate calculations.
+func (l *LSTM) gateCalculation(
+	Xt, W, Wb, H, R, Rb, P, C tensor.Tensor, activation ops.Activation,
+) (tensor.Tensor, error) {
+	gemm := &Gemm{transA: false, transB: true, alpha: 1.0, beta: 1.0}
+
+	inputCalc, err := gemm.Apply([]tensor.Tensor{Xt, W, Wb})
+	if err != nil {
+		return nil, err
+	}
+
+	hiddenCalc, err := gemm.Apply([]tensor.Tensor{H, R, Rb})
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := tensor.Add(inputCalc[0], hiddenCalc[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if P != nil {
+		C, broadcastedP, err := ops.UnidirectionalBroadcast(C, P)
+		if err != nil {
+			return nil, err
+		}
+
+		peepholeActivation, err := tensor.Mul(broadcastedP, C)
+		if err != nil {
+			return nil, err
+		}
+
+		output, err = tensor.Add(output, peepholeActivation)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return activation(output)
+}
+
+// cellCalculation performs the calculation of the LSTM cell update defined by:
+//
+//	Ct = ft (.) Ct-1 + it (.) ct
+//
+// Where 'ft' is the forget gate activation at time t, (.) denotes element-wise
+// multiplication, 'Ct-1' denotes the cell state at time t-1, 'it' denotes the input
+// gate activation at time t and 'ct' denotes the cell state activation at time t (which)
+// is not the same as Ct or Ct-1).
+func (l *LSTM) cellCalculation(ft, it, ct, Ct tensor.Tensor) (tensor.Tensor, error) {
+	cellForget, err := tensor.Mul(ft, Ct)
+	if err != nil {
+		return nil, err
+	}
+
+	cellInput, err := tensor.Mul(it, ct)
+	if err != nil {
+		return nil, err
+	}
+
+	return tensor.Add(cellForget, cellInput)
+}
+
+// hiddenCalculation performs the calculation of the new LSTM hidden state defined by:
+//
+//	Ht = ot (.) h(Ct)
+//
+// Where Ht is the new hidden state at time t, 'ot' is the output at time t, (.) denotes
+// element-wise multiplication, 'h()' denotes an activation function and 'Ct' denotes the
+// cell state at time t.
+func (l *LSTM) hiddenCalculation(ot, Ct tensor.Tensor, activation ops.Activation) (tensor.Tensor, error) {
+	cellActivated, err := activation(Ct)
+	if err != nil {
+		return nil, err
+	}
+
+	return tensor.Mul(ot, cellActivated)
+}
+
+// getWeights splits tensor W into 4 weight matrices.
+func (l *LSTM) getWeights(W tensor.Tensor) (Wi, Wo, Wf, Wh tensor.Tensor, err error) {
+	nWeightMatrices := 4
+	nWeightDimensions := 3
```

Comment:
Are these defined by the standard? why 4 and 3? maybe a comment? And if they are defined than maybe define a const for them?